include/numerics/petsc_shell_matrix.h

```diff
@@ -0,0 +1,203 @@
+// The libMesh Finite Element Library.
+// Copyright (C) 2002-2019 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
+#ifndef LIBMESH_PETSC_SHELL_MATRIX_H
+#define LIBMESH_PETSC_SHELL_MATRIX_H
+
+
+#include "libmesh/libmesh_config.h"
+#ifdef LIBMESH_HAVE_PETSC
+
+
+// Local includes
+#include "libmesh/libmesh_common.h"
+#include "libmesh/reference_counted_object.h"
+#include "libmesh/libmesh.h"
+#include "libmesh/shell_matrix.h"
+#include "libmesh/petsc_macro.h"
+#include "libmesh/petsc_solver_exception.h"
+#include "libmesh/petsc_vector.h"
+#include "libmesh/libmesh_common.h"
+
+// Petsc include files.
+#ifdef I
+# define LIBMESH_SAW_I
+#endif
+#include <petscmat.h>
+#ifndef LIBMESH_SAW_I
+# undef I // Avoid complex.h contamination
+#endif
+
+namespace libMesh
+{
+
+/**
+ * This class allows to use a PETSc shell matrix.
+ * All overridden virtual functions are documented in
+ * shell_matrix.h.
+ *
+ * \author Fande Kong (fdkong.jd@gmail.com)
+ * \date 2019
+ */
+template <typename T>
+class PetscShellMatrix : public ShellMatrix<T>
```

Comment:
What's the advantage of this over using a SparseShellMatrix with a PetscMatrix?

Reply:
`PetscMatrix` suppose to work with nonshell matrices, e.g., SeqAIJ, MPIAIJ.

`PetscShellMatrix` is partially attached to shell matrix. Shell matrix does not need a preallocation, but needs to attach constimized `MatVec`.

BTW, I do not completely understand how to use `a SparseShellMatrix with a PetscMatrix` 
to achieve what I have done here. 

It may help you see more if you would like to take a look at a use case here https://github.com/idaholab/moose/pull/14406

I am open to any other options, if you have other thoughts on how to achieve my goal.

Reply:
> PetscMatrix suppose to work with nonshell matrices, e.g., SeqAIJ, MPIAIJ.

How hard would it be to remove that requirement, though?  I see we have a PetscMatrix constructor that takes an externally-managed Mat object, and I don't see any internally-used methods except init() where it enforces the type of matrix.  If we can replace 300 lines of PetscShellMatrix code with 30 more lines of switch(_mat_type) cases in PetscMatrix::init(), that'd be worth it.

Reply:
Do you mean `PetscMatrix` take a shell mat of PETSc?

The big concern is that almost all the APIs in PetscMatrix will be invalid for` shell matrix`. For example, the following APIs will no sense for shell matrix:


  ```
virtual void set (const numeric_index_type i,
                    const numeric_index_type j,
                    const T value) override;

  virtual void add (const numeric_index_type i,
                    const numeric_index_type j,
                    const T value) override;

  virtual void add_matrix (const DenseMatrix<T> & dm,
                           const std::vector<numeric_index_type> & rows,
                           const std::vector<numeric_index_type> & cols) override;

  virtual void add_matrix (const DenseMatrix<T> & dm,
                           const std::vector<numeric_index_type> & dof_indices) override;

  virtual void add_block_matrix (const DenseMatrix<T> & dm,
                                 const std::vector<numeric_index_type> & brows,
                                 const std::vector<numeric_index_type> & bcols) override;

  virtual void add_block_matrix (const DenseMatrix<T> & dm,
                                 const std::vector<numeric_index_type> & dof_indices) override
  { this->add_block_matrix (dm, dof_indices, dof_indices); }

  /**
   * Compute A += a*X for scalar \p a, matrix \p X.
   *
   * \note The matrices A and X need to have the same nonzero pattern,
   * otherwise \p PETSc will crash!
   *
   * \note It is advisable to not only allocate appropriate memory
   * with \p init(), but also explicitly zero the terms of \p this
   * whenever you add a non-zero value to \p X.
   *
   * \note \p X will be closed, if not already done, before performing
   * any work.
   */
  virtual void add (const T a, const SparseMatrix<T> & X) override;

  virtual T operator () (const numeric_index_type i,
                         const numeric_index_type j) const override;

  virtual Real l1_norm () const override;

  virtual Real linfty_norm () const override;

```

I do not think it will be better than the current implementation.

Reply:
Similarly, we did not merge `ShellMatrix` into `SparseMatrix`

Reply:
Yeah, good point.

In hindsight perhaps we should have made something like ShellMatrix into a SparseMatrix superclass, though, rather than relying on overload after overload for the cases where we only need the former's capabilities but we want to handle the latter too.